File: frontend/src/types/cash-flow.ts
// --- Enum types ---

export type ForecastStatus = "bozza" | "attiva" | "archiviata";
export type ForecastType = "base" | "ottimistico" | "pessimistico" | "personalizzato";
export type CashFlowLineType = "entrata" | "uscita";
export type CashFlowSource =
  | "movimento_reale"
  | "scadenza"
  | "pagamento_pianificato"
  | "ricorrenza"
  | "manuale"
  | "stima_storica";
export type ConfidenceLevel = "certa" | "alta" | "media" | "bassa";
export type ScenarioAdjustmentType = "aggiunta" | "rimozione" | "modifica_importo" | "spostamento_data";
export type CashFlowAlertType = "sotto_soglia_minima" | "sopra_soglia_massima" | "saldo_negativo" | "varianza_alta";
export type AlertStatus = "attivo" | "risolto" | "ignorato";

// --- Label maps ---

export const FORECAST_STATUS_LABELS: Record<ForecastStatus, string> = {
  bozza: "Bozza",
  attiva: "Attiva",
  archiviata: "Archiviata",
};

export const FORECAST_TYPE_LABELS: Record<ForecastType, string> = {
  base: "Base",
  ottimistico: "Ottimistico",
  pessimistico: "Pessimistico",
  personalizzato: "Personalizzato",
};

export const CASH_FLOW_SOURCE_LABELS: Record<CashFlowSource, string> = {
  movimento_reale: "Movimento Reale",
  scadenza: "Scadenza",
  pagamento_pianificato: "Pagamento Pianificato",
  ricorrenza: "Ricorrenza",
  manuale: "Manuale",
  stima_storica: "Stima Storica",
};

export const CONFIDENCE_LABELS: Record<ConfidenceLevel, string> = {
  certa: "Certa",
  alta: "Alta",
  media: "Media",
  bassa: "Bassa",
};

export const ALERT_TYPE_LABELS: Record<CashFlowAlertType, string> = {
  sotto_soglia_minima: "Sotto Soglia Minima",
  sopra_soglia_massima: "Sopra Soglia Massima",
  saldo_negativo: "Saldo Negativo",
  varianza_alta: "Varianza Alta",
};

export const ALERT_STATUS_LABELS: Record<AlertStatus, string> = {
  attivo: "Attivo",
  risolto: "Risolto",
  ignorato: "Ignorato",
};

export const SCENARIO_TYPE_LABELS: Record<ScenarioAdjustmentType, string> = {
  aggiunta: "Aggiunta",
  rimozione: "Rimozione",
  modifica_importo: "Modifica Importo",
  spostamento_data: "Spostamento Data",
};

// --- Status colors ---

export const FORECAST_STATUS_COLORS: Record<ForecastStatus, string> = {
  bozza: "bg-[#f2f2f2] text-[#817f7d]",
  attiva: "bg-[#8cddba] text-[#1a1a1a]",
  archiviata: "bg-[#c0c8ff] text-[#1a1a1a]",
};

export const CONFIDENCE_COLORS: Record<ConfidenceLevel, string> = {
  certa: "bg-[#8cddba] text-[#1a1a1a]",
  alta: "bg-[#c0c8ff] text-[#1a1a1a]",
  media: "bg-[#ffd98c] text-[#1a1a1a]",
  bassa: "bg-[#e5e4e3] text-[#817f7d]",
};

export const ALERT_STATUS_COLORS: Record<AlertStatus, string> = {
  attivo: "bg-[#fddcdc] text-[#ea0b0b]",
  risolto: "bg-[#8cddba] text-[#1a1a1a]",
  ignorato: "bg-[#e5e4e3] text-[#817f7d]",
};

export const ALERT_TYPE_COLORS: Record<CashFlowAlertType, string> = {
  sotto_soglia_minima: "bg-[#ffd98c] text-[#1a1a1a]",
  sopra_soglia_massima: "bg-[#c0c8ff] text-[#1a1a1a]",
  saldo_negativo: "bg-[#fddcdc] text-[#ea0b0b]",
  varianza_alta: "bg-[#c8b9df] text-[#1a1a1a]",
};

// --- Interfaces ---

export interface CashFlowProjectionLine {
  data: string;
  tipo: CashFlowLineType;
  importo: number;
  categoria: string | null;
  descrizione: string | null;
  fonte: CashFlowSource;
  fonte_id: string | null;
  confidenza: ConfidenceLevel;
  saldo_progressivo: number;
  is_realizzata: boolean;
}

export interface CashFlowProjection {
  saldo_iniziale: number;
  saldo_finale: number;
  totale_entrate: number;
  totale_uscite: number;
  data_inizio: string;
  data_fine: string;
  lines: CashFlowProjectionLine[];
}

export interface CashFlowSummary {
  saldo_attuale: number;
  previsione_7gg: number;
  previsione_30gg: number;
  previsione_90gg: number;
  trend_7gg: number;
  trend_30gg: number;
  prossimo_alert: CashFlowAlert | null;
  runway_mesi: number | null;
  burn_rate_mensile: number | null;
}

export interface CashFlowForecastLine {
  id: string;
  forecast_id: string;
  data: string;
  tipo: CashFlowLineType;
  importo: number;
  categoria: string | null;
  descrizione: string | null;
  fonte: CashFlowSource;
  fonte_id: string | null;
  confidenza: ConfidenceLevel;
  saldo_progressivo: number;
  is_realizzata: boolean;
  varianza: number | null;
  created_at: string;
  updated_at: string;
}

export interface CashFlowScenario {
  id: string;
  forecast_id: string;
  nome: string;
  descrizione: string | null;
  tipo_aggiustamento: ScenarioAdjustmentType;
  data_originale: string | null;
  data_modificata: string | null;
  importo_originale: number | null;
  importo_modificato: number | null;
  tipo_movimento: CashFlowLineType | null;
  categoria: string | null;
  attivo: boolean;
  created_at: string;
  updated_at: string;
}

export interface CashFlowAlert {
  id: string;
  company_id: string;
  forecast_id: string | null;
  tipo: CashFlowAlertType;
  data_prevista: string;
  saldo_previsto: number;
  soglia: number;
  messaggio: string;
  stato: AlertStatus;
  notificato: boolean;
  notificato_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface CashFlowForecast {
  id: string;
  company_id: string;
  creato_da_user_id: string | null;
  nome: string;
  descrizione: string | null;
  data_inizio: string;
  data_fine: string;
  saldo_iniziale: number;
  saldo_finale_previsto: number;
  totale_entrate_previste: number;
  totale_uscite_previste: number;
  stato: ForecastStatus;
  tipo: ForecastType;
  parametri: Record<string, unknown> | null;
  is_auto: boolean;
  created_at: string;
  updated_at: string;
}

export interface CashFlowForecastDetail extends CashFlowForecast {
  lines: CashFlowForecastLine[];
  scenarios: CashFlowScenario[];
}

export interface CashFlowForecastListResponse {
  items: CashFlowForecast[];
  total: number;
  page: number;
  page_size: number;
}

export interface CashFlowAlertListResponse {
  items: CashFlowAlert[];
  total: number;
}

export interface VarianceLine {
  data: string;
  categoria: string | null;
  importo_previsto: number;
  importo_reale: number;
  varianza: number;
  varianza_percentuale: number | null;
}

export interface VarianceResponse {
  forecast_id: string;
  accuracy: number;
  totale_previsto: number;
  totale_reale: number;
  varianza_totale: number;
  lines: VarianceLine[];
}

export interface ScenarioCompare {
  base: CashFlowProjection;
  con_scenari: CashFlowProjection;
  delta_saldo_finale: number;
}

export interface SeasonalityMonth {
  mese: number;
  entrate_medie: number;
  uscite_medie: number;
  saldo_netto_medio: number;
}

export interface SeasonalityResponse {
  mesi: SeasonalityMonth[];
  periodo_analisi_mesi: number;
}

export interface TrendPoint {
  periodo: string;
  entrate: number;
  uscite: number;
  saldo_netto: number;
}

export interface TrendResponse {
  punti: TrendPoint[];
  trend_entrate: string;
  trend_uscite: string;
  raggruppamento: string;
}

export interface RunwayResponse {
  saldo_attuale: number;
  burn_rate_mensile: number;
  runway_mesi: number;
  data_esaurimento: string | null;
  mesi_analisi: number;
}

export interface CashFlowChartPoint {
  periodo: string;
  saldo: number;
  entrate: number;
  uscite: number;
}

export interface CashFlowChartResponse {
  punti: CashFlowChartPoint[];
  raggruppamento: string;
}

export interface CashFlowTableRow {
  categoria: string;
  tipo: CashFlowLineType;
  periodi: Record<string, number>;
  totale: number;
}

export interface CashFlowTableResponse {
  righe: CashFlowTableRow[];
  periodi: string[];
  totali_entrate: Record<string, number>;
  totali_uscite: Record<string, number>;
  saldi_netti: Record<string, number>;
}

File: backend/app/schemas/cash_flow.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Response schemas ---

class CashFlowForecastLineResponse(BaseModel):
    id: uuid.UUID
    forecast_id: uuid.UUID
    data: date
    tipo: str
    importo: Decimal
    categoria: str | None = None
    descrizione: str | None = None
    fonte: str
    fonte_id: str | None = None
    confidenza: str
    saldo_progressivo: Decimal
    is_realizzata: bool
    varianza: Decimal | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CashFlowScenarioResponse(BaseModel):
    id: uuid.UUID
    forecast_id: uuid.UUID
    nome: str
    descrizione: str | None = None
    tipo_aggiustamento: str
    data_originale: date | None = None
    data_modificata: date | None = None
    importo_originale: Decimal | None = None
    importo_modificato: Decimal | None = None
    tipo_movimento: str | None = None
    categoria: str | None = None
    attivo: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CashFlowAlertResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    forecast_id: uuid.UUID | None = None
    tipo: str
    data_prevista: date
    saldo_previsto: Decimal
    soglia: Decimal
    messaggio: str
    stato: str
    notificato: bool
    notificato_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CashFlowForecastResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    creato_da_user_id: uuid.UUID | None = None
    nome: str
    descrizione: str | None = None
    data_inizio: date
    data_fine: date
    saldo_iniziale: Decimal
    saldo_finale_previsto: Decimal
    totale_entrate_previste: Decimal
    totale_uscite_previste: Decimal
    stato: str
    tipo: str
    parametri: dict | None = None
    is_auto: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CashFlowForecastDetailResponse(CashFlowForecastResponse):
    lines: list[CashFlowForecastLineResponse] = []
    scenarios: list[CashFlowScenarioResponse] = []


class CashFlowForecastListResponse(BaseModel):
    items: list[CashFlowForecastResponse]
    total: int
    page: int
    page_size: int


# --- Real-time projection response ---

class CashFlowProjectionLine(BaseModel):
    data: date
    tipo: str
    importo: Decimal
    categoria: str | None = None
    descrizione: str | None = None
    fonte: str
    fonte_id: str | None = None
    confidenza: str
    saldo_progressivo: Decimal
    is_realizzata: bool = False


class CashFlowProjectionResponse(BaseModel):
    saldo_iniziale: Decimal
    saldo_finale: Decimal
    totale_entrate: Decimal
    totale_uscite: Decimal
    data_inizio: date
    data_fine: date
    lines: list[CashFlowProjectionLine]


class CashFlowSummaryResponse(BaseModel):
    saldo_attuale: Decimal
    previsione_7gg: Decimal
    previsione_30gg: Decimal
    previsione_90gg: Decimal
    trend_7gg: Decimal
    trend_30gg: Decimal
    prossimo_alert: CashFlowAlertResponse | None = None
    runway_mesi: Decimal | None = None
    burn_rate_mensile: Decimal | None = None


# --- Variance ---

class VarianceLineResponse(BaseModel):
    data: date
    categoria: str | None = None
    importo_previsto: Decimal
    importo_reale: Decimal
    varianza: Decimal
    varianza_percentuale: Decimal | None = None


class VarianceResponse(BaseModel):
    forecast_id: uuid.UUID
    accuracy: Decimal
    totale_previsto: Decimal
    totale_reale: Decimal
    varianza_totale: Decimal
    lines: list[VarianceLineResponse]


# --- Scenario compare ---

class ScenarioCompareResponse(BaseModel):
    base: CashFlowProjectionResponse
    con_scenari: CashFlowProjectionResponse
    delta_saldo_finale: Decimal


# --- Analysis ---

class SeasonalityMonthResponse(BaseModel):
    mese: int
    entrate_medie: Decimal
    uscite_medie: Decimal
    saldo_netto_medio: Decimal


class SeasonalityResponse(BaseModel):
    mesi: list[SeasonalityMonthResponse]
    periodo_analisi_mesi: int


class TrendPointResponse(BaseModel):
    periodo: str
    entrate: Decimal
    uscite: Decimal
    saldo_netto: Decimal


class TrendResponse(BaseModel):
    punti: list[TrendPointResponse]
    trend_entrate: str  # crescente, decrescente, stabile
    trend_uscite: str
    raggruppamento: str


class RunwayResponse(BaseModel):
    saldo_attuale: Decimal
    burn_rate_mensile: Decimal
    runway_mesi: Decimal
    data_esaurimento: date | None = None
    mesi_analisi: int


# --- Chart/Table (PRD-08) ---

class CashFlowChartPoint(BaseModel):
    periodo: str
    saldo: Decimal
    entrate: Decimal
    uscite: Decimal


class CashFlowChartResponse(BaseModel):
    punti: list[CashFlowChartPoint]
    raggruppamento: str


class CashFlowTableRow(BaseModel):
    categoria: str
    tipo: str  # entrata / uscita
    periodi: dict[str, Decimal]
    totale: Decimal


class CashFlowTableResponse(BaseModel):
    righe: list[CashFlowTableRow]
    periodi: list[str]
    totali_entrate: dict[str, Decimal]
    totali_uscite: dict[str, Decimal]
    saldi_netti: dict[str, Decimal]


class CashFlowCellDetailItem(BaseModel):
    data: date
    descrizione: str | None = None
    importo: Decimal
    fonte: str
    controparte: str | None = None


class CashFlowCellDetailResponse(BaseModel):
    categoria: str
    periodo: str
    tipo: str
    items: list[CashFlowCellDetailItem]
    totale: Decimal


# --- Request schemas ---

class CashFlowForecastCreateRequest(BaseModel):
    nome: str = Field(min_length=1, max_length=255)
    descrizione: str | None = None
    data_inizio: date
    data_fine: date
    tipo: str = "base"
    bank_account_ids: list[uuid.UUID] | None = None
    includi_scadenze: bool = True
    includi_pagamenti: bool = True
    includi_ricorrenze: bool = True
    includi_stime_storiche: bool = False


class CashFlowForecastUpdateRequest(BaseModel):
    nome: str | None = Field(None, min_length=1, max_length=255)
    descrizione: str | None = None
    stato: str | None = None


class CashFlowScenarioCreateRequest(BaseModel):
    nome: str = Field(min_length=1, max_length=255)
    descrizione: str | None = None
    tipo_aggiustamento: str
    data_originale: date | None = None
    data_modificata: date | None = None
    importo_originale: Decimal | None = None
    importo_modificato: Decimal | None = None
    tipo_movimento: str | None = None
    categoria: str | None = None


class CashFlowScenarioUpdateRequest(BaseModel):
    nome: str | None = Field(None, min_length=1, max_length=255)
    descrizione: str | None = None
    tipo_aggiustamento: str | None = None
    data_originale: date | None = None
    data_modificata: date | None = None
    importo_originale: Decimal | None = None
    importo_modificato: Decimal | None = None
    tipo_movimento: str | None = None
    categoria: str | None = None
    attivo: bool | None = None


# --- Budget (PRD-08) ---

class BudgetCreateRequest(BaseModel):
    categoria: str
    tipo: str  # entrata / uscita
    importi_mensili: dict[str, Decimal]  # {"2026-01": 5000, "2026-02": 4500}


class BudgetSuggestionResponse(BaseModel):
    categoria: str
    tipo: str
    importo_suggerito: Decimal
    base_calcolo: str  # media_3_mesi, media_6_mesi, etc.


# --- Alert list response ---

class CashFlowAlertListResponse(BaseModel):
    items: list[CashFlowAlertResponse]
    total: int
